runner_service/inventory.py

```diff
@@ -91,37 +99,52 @@ def load(self):
 
         if not os.path.exists(self.filename):
             try:
-                with open(self.filename, 'w') as inv:
+                # Using Python 3 exclusive creation
+                with open(self.filename, 'x') as inv:                    
                     inv.write(yaml.dump(AnsibleInventory.inventory_seed,
-                                        default_flow_style=False))
+                                            default_flow_style=False))
+            except FileExistsError:
+                logger.info("Inventory file '{}' already created".format(self.filename))
             except IOError:
                 raise InventoryWriteError("Unable to create the seed inventory"
                                           " file at {}".format(self.filename))
-
-        if self.exclusive_lock:
-
-            try:
-                self.fd = open(self.filename, 'r+')
-                self.lock()
-            except (BlockingIOError, OSError):
-                # Can't obtain an exclusive_lock
-                self.fd.close()
-                return
+            
+        try:    
+            if self.exclusive_lock:
+                try:
+                    self.fd = open(self.filename, 'r+')
+                    self.lock()
+                except (BlockingIOError, OSError):
+                    # Can't obtain an exclusive_lock
+                    self.fd.close()
+                    return
+                else:
+                    raw = self.fd.read().strip()
             else:
-                raw = self.fd.read().strip()
-        else:
-            raw = fread(self.filename)
+                raw = fread(self.filename)
+        except Exception as ex:
+            raise InventoryreadError("Unable to read the inventory"
+                                     " file at {}, error: {}".format(self.filename, ex))
 
         if not raw:
-            self.inventory = None
+            # If the inventory is empty for some extrange reason
+            self.inventory = AnsibleInventory.inventory_seed
         else:
-            # TODO what happens with invalid yaml?
-            self.inventory = yaml.safe_load(raw)
-
+            # invalid yaml management
+            try:
+                self.inventory = yaml.safe_load(raw)
+            except yaml.YAMLError as ex:
+                raise InventoryCorruptError("Unable to understand the inventory"
+                                            " yaml file at {}, error: {}".format(self.filename, ex))
+        
     def _dump(self):
         return yaml.dump(self.inventory, default_flow_style=False)
 
     def save(self):
+        # Get the only when fd when we are going to save
+        self.fd = open(self.filename, 'w')
```

Comment:
why do we need to do this here? When the inventory is loaded and the exclusive lock is requested, the fs is locked until we exit/save - so not sure what the scenario is that this addresses?

Reply:
This is the behaviour when you cretate the inventory objects with: AnsibleInventory(excl=True)
If you do not set especifically excl=True, the fd attribute of the object is not initialized. and all the calls to the "save" method return error because the "seek" method can not be used in a "none" object.

I not sure about what do you wanted exactly...maybe to provide a "read" only inventory when the "excl"  parameter is not set to True? if it is the aim then the "save" method must be disabled  when excl=False.

I made this modification because in the unit tests the inventory is created with "Excl=false" and write operations are executed ( in fact all  the "write" test failed because fd not initialized).

Tell me what is what do you want/prefer  and i will adjust the code. :-)

Reply:
I think I wrote the test cases prior to adding the exclusive lock parameter. The code only uses excl for update calls - GET calls don't use it and don't need to lock.

The inventory object is not created upon each call. For a GET request the object doesn't lock, but when we said add_group - we create the inventory object and lock the underlying file to prevent other writers. the 'save' method is only invoked for update methods anyway.

I think the problem that we have is that the unittest needs to be restructured - perhaps to create the inventory object within each testcase to open it correctly

Reply:
Ok. I understand now.  I will fix the unit tests and restore the code to the previous state ( improving it if that is possible :-)